PFR/Entropy/Basic.lean

```diff
@@ -306,16 +306,120 @@ lemma entropy_eq_log_card {X : Ω → S} (hX : Measurable X) (μ : Measure Ω) (
   rw [entropy_def, measureEntropy_eq_card_iff_measure_eq, Measure.map_apply hX MeasurableSet.univ]
   simp
 
+lemma Finset.Nonempty_image {α β : Type*} [DecidableEq β]
+{s : Finset α} (hs : s.Nonempty) (f : α → β) : (s.image f).Nonempty := by
+  rw [Finset.Nonempty] at hs ⊢
+  obtain ⟨x, hx⟩ := hs
+  use (f x)
+  exact Finset.mem_image_of_mem f hx
+
+lemma Finset.image_min'_le_sum {α β : Type*} [DecidableEq β] [LinearOrder β] [AddCommMonoid β]
+{s : Finset α} (hs : s.Nonempty) (f : α → β) (hf : ∀ a, 0 ≤ f a ):
+  (s.image f).min' (Finset.Nonempty_image hs f) ≤ s.sum f := sorry
+
+lemma Measure.eq_zero_of_map_eq_zero (X : Ω → S) (μ : Measure Ω) (hX : AEMeasurable X μ)
+  (hμ : μ.map X = 0) : μ = 0 := by
+  have := Measure.preimage_null_of_map_null hX (Measure.measure_univ_eq_zero.mpr hμ)
+  rwa [preimage_univ, Measure.measure_univ_eq_zero] at this
+
+def Measure.NeZero_map_NeZero {X : Ω → S} (μ : Measure Ω) (hX : AEMeasurable X μ)
+  [NeZero μ] : NeZero (μ.map X) := sorry
+
+lemma IsFiniteMeasure_sum_measure_singleton [Nonempty S] (μ : Measure S)
+  (hμ' : IsFiniteMeasure μ) [NeZero μ] : ∑ s : S, μ {s} = μ Set.univ :=
+  sorry
+
+lemma IsFiniteMeasure_sum_measureReal_singleton [Nonempty S] (μ : Measure S)
+  [IsFiniteMeasure μ] [NeZero μ] : ∑ s : S, μ.real {s} = μ.real Set.univ := sorry
+
+lemma IsFiniteMeasure_sum_measure_univ_inv_mul_measure_singleton [Nonempty S] (μ : Measure S)
+  [IsFiniteMeasure μ] [NeZero μ] : ∑ s : S, ((μ Set.univ)⁻¹ * μ {s}) = 1:= sorry
+
+lemma IsFiniteMeasure_sum_measureReal_univ_inv_mul_measureReal_singleton [Nonempty S] (μ : Measure S)
+  [IsFiniteMeasure μ] [NeZero μ] : ∑ s : S, ((μ.real Set.univ)⁻¹ * μ.real {s}) = 1:= sorry
+
 /-- If $X$ is an $S$-valued random variable, then there exists $s \in S$ such that
 $P[X=s] \geq \exp(-H[X])$. -/
-lemma prob_ge_exp_neg_entropy (X : Ω → S) (μ : Measure Ω) :
-  ∃ s : S, μ.map X {s} ≥ (μ Set.univ) * (rexp (- H[X ; μ])).toNNReal := by sorry
+lemma prob_ge_exp_neg_entropy [Nonempty S] (X : Ω → S) (hX : Measurable X) (μ : Measure Ω)
+  [hμ' : IsFiniteMeasure μ] [NeZero μ] (hbd : ∀ (s : S), μ.map X {s} ≠ 0) :
```

Comment:
are we allowed to add these assumptions?
- `Nonempty S` is already added via `variables` in this file
- we can derive `hμ'` and `NeZero \mu` via `rcases` (the statement degenerates if either is false)
- not sure about `hX` and `hbd`. IIRC i didn't need `hX`.

Reply:
What I ended up doing for `hbd` is to use `Finset.filter` to specify only summing nonzero values; not sure if it was the most efficient way though

Reply:
A lot of the assumptions can and should be removed (I added them as I went as was planning to remove them in the golfing process :)). 
 For `hbd`, I guess in my proof too I could be removed by writing H as a sum over the elements with non zero probability (maybe this fact should be stated in a separate lemmas?)